crypto/rand/rand_lib.c

```diff
@@ -42,134 +42,126 @@ int rand_fork_count;
-        CRYPTO_THREAD_write_lock(rand_bytes.lock);
-    }
-
-    /* Get desired amount, but no more than we have. */
-    if (min_len > rand_bytes.curr)
-        min_len = rand_bytes.curr;
-    if (min_len != 0) {
-        memcpy(randomness, rand_bytes.buff, min_len);
-        /* Update amount left and shift it down. */
-        rand_bytes.curr -= min_len;
-        if (rand_bytes.curr != 0)
-            memmove(rand_bytes.buff, &rand_bytes.buff[min_len], rand_bytes.curr);
-    }
-    CRYPTO_THREAD_unlock(rand_bytes.lock);
-    *pout = randomness;
-    return min_len;
-}
 
-size_t drbg_entropy_from_parent(RAND_DRBG *drbg,
-                                unsigned char **pout,
-                                int entropy, size_t min_len, size_t max_len)
+/*
+ * Implements the get_entropy() callback
+ *
+ * If the DRBG has a parent, then the required amount of entropy input
+ * is fetched using the parent's RAND_DRBG_generate().
+ *
+ * Otherwise, the entropy is polled from the system entropy sources.
+ * Previously, this used to be done by calling RAND_poll().
+ * Nowadays, both RAND_poll() and drbg_get_entropy() are based on 
+ * the RAND_POOL API, i.e., the entropy is aqcuired by RAND_POOL_fill().
+ */
+size_t drbg_get_entropy(RAND_DRBG *drbg,
+                        unsigned char **pout,
+                        int entropy, size_t min_len, size_t max_len)
 {
-    int st;
-    unsigned char *randomness;
+    size_t ret = 0;
+    int entropy_available = 0;
+
+    RAND_POOL* pool = RAND_POOL_new(entropy, min_len, max_len);
+    
+    if (drbg->parent) {
+        int bytes_needed = RAND_POOL_bytes_needed(pool, 8);
+        unsigned char *buffer = RAND_POOL_add_begin(pool, bytes_needed);
+
+        if (buffer != NULL) {
+            /* Get entropy from parent, include our state as additional input. */
+            int bytes = 0;
+
+            if (RAND_DRBG_generate(drbg->parent,
+                                    buffer, bytes_needed,
+                                    0,
+                                    (unsigned char *)drbg, sizeof(*drbg)) != 0)
+                bytes = bytes_needed;
+            
+            entropy_available = RAND_POOL_add_end(pool, bytes, bytes);
+        }
 
-    if (min_len > (size_t)drbg->size) {
-        /* Should not happen.  See comment near RANDOMNESS_NEEDED. */
-        min_len = drbg->size;
+    } else {
+        /* Get entropy by polling system entropy sources. */
+        entropy_available = RAND_POOL_fill(pool);
     }
 
-    randomness = drbg->secure ? OPENSSL_secure_malloc(drbg->size)
-                                    : OPENSSL_malloc(drbg->size);
-
-    /* Get random from parent, include our state as additional input. */
-    st = RAND_DRBG_generate(drbg->parent, randomness, min_len, 0,
-                            (unsigned char *)drbg, sizeof(*drbg));
-    if (st == 0) {
-        drbg_release_entropy(drbg, randomness, min_len);
-        return 0;
+    if (entropy_available > 0) {
```

Comment:
At first sight, this looks wrong. entropy_available should be >= entropy (not sure about bits vs bytes). Only if that's the case should it return success.